### PR TITLE
V1.4

### PR DESCRIPTION
**Pull request by kfiroo**
Hey all,
This is a major release that has a lot of breaking changes.
All changes came from community requests, I took all the PRs and Issues discussions and compiled them into this new version.
If anyone has anything to add/comment this is the place.

* Move all cache logic to `ImageCacheManager`
* Introduce `ImageCacheProvider` - a top-level component that provides an entry point for customization over the `ImageCacheManager`
* Separate concerns of `cache` and `storage`.
  * `MemoryCache` manages the caching of URLs, this means that for each URL there is an entry in the cache with its expiration date, local file path, etc. - all decisions are made based on the url cache layer.
  * fs layer has no knowledge of the cache layer and is orchestrated by the `ImageCacheManager`
* Use ES6 classes instead of the deprecated `React.createClass`
* Use `prop-types` instead of the deprecated `React.PropTypes`

### CODE REVIEW

**Comment by kfiroo** [comment]
@iegik @alinz @markusguenther @pablocarrillo @markusguenther @ovaris
Hey guys,
Please see this PR and let me know what you think :)

File: package.json [hunk 1]
```diff
@@ -24,12 +24,19 @@
...
+    "utils/fsUtils.js",
+    "utils/pathUtils.js"
   ],
   "dependencies": {
     "crypto-js": "3.1.9-1",
     "lodash": "4.17.4",
-    "react-native-fetch-blob": "0.10.4",
+    "prop-types": "15.5.10",
+    "react-native-clcasher": "1.0.0",
+    "react-native-fetch-blob": "0.10.7",
```
Comment (alinz) [inline package.json]: react-native-fetch-blob needs to be set to 0.10.8. The problem described here https://github.com/wkh237/react-native-fetch-blob/pull/456. and it has been fixed in 0.10.8

**Comment by alinz** [comment]
@kfiroo Thanks for this code. One other thing. react-native has been upgraded to 0.47. I hurts. As soon as you upgrade, they upgrade. What I'm suggesting is release this one and release the next version updating to react-native 0.47.

Make sure to update doc.

Awesome job!

**Comment by iegik** [comment]
I will make a review, as soon, as I back to mobile app. Thanks for PR

**Comment by alinz** [comment]
@kfiroo can you release `1.4` as `1.4-rc` and publish it to npm so I can pull it and testing it against my app. it would be a lot safer for everyone else if you have a apis breaking.

**Comment by markusguenther** [comment]
@kfiroo Will try it soon, but code wise it looks good 👍

**Comment by kfiroo** [comment]
@alinz can you explain again what you said about RN47?

Version `1.4.0-rc` is published to npm under the next tag so
`npm install react-native-cached-image@next` and `npm install react-native-cached-image@1.4.0-rc` will install the latest version

**Comment by alinz** [comment]
@kfiroo haha, I tried to read what I wrote. I couldn't understand what I have written. such an embarrassment. 

I think I was going to say upgrade the react-native in example folder. Which is not that important at the moment. Sorry for confusion.
 
I will try the rc against couple of my apps and get back to you.

Thanks

**Comment by kfiroo** [comment]
Hey,
I updated the README so I'd appreciate some feedback on that too :)

**Comment by pablocarrillo** [comment]
Code looks good to me, I will test this week and see if the limit for concurrent downloads works as I need

**Comment by kytwb** [comment]
Seems like we would want this PR to be merged first so we can move forward with the other issues / PRs?

**Comment by iegik** [comment]
Don't forget create 1.3 branch

**Comment by kfiroo** [comment]
Hey,
Unless there are any objections I will promote this RC tomorrow :)

**Comment by kytwb** [comment]
@kfiroo any news?

**Comment by kfiroo** [comment]
Thanks for your patience all!

`v1.3` branch created from master.
`v1.4` merged into master.
version `1.4.0` published to npm.

May god help us all 😄

**Pull request merged**